stan/math/rev/functor/coupled_ode_system.hpp

```diff
@@ -438,53 +112,62 @@ struct coupled_ode_system<F, var, var> {
    *    expected number of derivatives, N.
    */
   void operator()(const std::vector<double>& z, std::vector<double>& dz_dt,
-                  double t) const {
+                  double t) {
     using std::vector;
 
+    dz_dt.resize(size());
+
     // Run nested autodiff in this scope
     nested_rev_autodiff nested;
 
-    vector<var> y_vars(z.begin(), z.begin() + N_);
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y_vars(N_);
+    for (size_t n = 0; n < N_; ++n)
+      y_vars.coeffRef(n) = z[n];
 
-    vector<var> dy_dt_vars = f_(t, y_vars, theta_nochain_, x_, x_int_, msgs_);
+    Eigen::Matrix<var, Eigen::Dynamic, 1> f_y_t_vars
+        = apply([&](auto&&... args) { return f_(t, y_vars, msgs_, args...); },
+                local_args_tuple_);
 
-    check_size_match("coupled_ode_system", "dz_dt", dy_dt_vars.size(), "states",
+    check_size_match("coupled_ode_system", "dy_dt", f_y_t_vars.size(), "states",
                      N_);
 
-    for (size_t i = 0; i < N_; i++) {
-      dz_dt[i] = dy_dt_vars[i].val();
-      dy_dt_vars[i].grad();
+    for (size_t i = 0; i < N_; ++i) {
+      dz_dt[i] = f_y_t_vars.coeffRef(i).val();
+      f_y_t_vars.coeffRef(i).grad();
 
-      for (size_t j = 0; j < N_; j++) {
-        // orders derivatives by equation (i.e. if there are 2 eqns
-        // (y1, y2) and 2 parameters (a, b), dy_dt will be ordered as:
-        // dy1_dt, dy2_dt, dy1_da, dy2_da, dy1_db, dy2_db
-        double temp_deriv = 0;
-        const size_t offset = N_ + N_ * j;
-        for (size_t k = 0; k < N_; k++) {
-          temp_deriv += z[offset + k] * y_vars[k].adj();
-        }
+      y_adjoints_ = y_vars.adj();
 
-        dz_dt[offset + i] = temp_deriv;
+      memset(args_adjoints_.data(), 0, sizeof(double) * num_args_vars);
+      apply(
+          [&](auto&&... args) {
+            accumulate_adjoints(args_adjoints_.data(), args...);
+          },
+          local_args_tuple_);
```

Comment:
I think instead of apply we should be using something like `for_each` here. Then we also wouldn't need those functions to all be recursive

https://gist.github.com/SteveBronder/410f7127f8be9dced2b1915765fd8644

Reply:
Not quite cause `save_varis` still needs to pass along the pointer it's saving things too (same with accumulate adjoints), but maybe there's a way to work around this too.

Reply:
ack, yeah we need to move up the pointer position as we accumulate